mipengine/controller/algorithm_executor_node_data_objects.py

```diff
@@ -0,0 +1,355 @@
+        self._node_id = full_name_split[1]
+        self._context_id = full_name_split[2]
+        self._command_id = full_name_split[3]
+        self._command_subid = full_name_split[4]
+
+    @property
+    def full_table_name(self) -> str:
+        return self._full_name
+
+    @property
+    def table_type(self) -> str:
+        return self._table_type
+
+    @property
+    def node_id(self) -> str:
+        return self._node_id
+
+    @property
+    def context_id(self) -> str:
+        return self._context_id
+
+    @property
+    def command_id(self) -> str:
+        return self._command_id
+
+    @property
+    def command_subid(self) -> str:
+        return self._command_subid
+
+    def without_node_id(self) -> str:
+        return (
+            self._table_type
+            + "_"
+            + self._context_id
+            + "_"
+            + self._command_id
+            + "_"
+            + self._command_subid
+        )
+
+    def __repr__(self):
+        return self.full_table_name
+
+
+class NodeSMPCTables(NodeData):
+    template: NodeTable
+    add_op: NodeTable
+    min_op: NodeTable
+    max_op: NodeTable
+    union_op: NodeTable
+
+    def __init__(self, template, add_op, min_op, max_op, union_op):
+        self.template = template
+        self.add_op = add_op
+        self.min_op = min_op
+        self.max_op = max_op
+        self.union_op = union_op
+
+
+class AlgoExecData(ABC):
+    """
+    AlgoExecData are representing one data object but could be located into
+    more than one node. For example the LocalNodesTable is treated as one table
+    from the algorithm executor but is located in multiple local nodes.
+    """
+
+    pass
+
+
+class LocalNodesData(AlgoExecData, ABC):
+    pass
+
+
+class GlobalNodeData(AlgoExecData, ABC):
+    pass
+
+
+# When _AlgorithmExecutionInterface::run_udf_on_local_nodes(..) is called, depending on
+# how many local nodes are participating in the current algorithm execution, several
+# database tables are created on all participating local nodes. Irrespectevely of the
+# number of local nodes participating, the number of tables created on each of these local
+# nodes will be the same.
+# Class _LocalNodeTable is the structure that represents the concept of these database
+# tables, created during the execution of a udf, in the algorithm execution layer. A key
+# concept is that a _LocalNodeTable stores 'pointers' to 'relevant' tables existing in
+# different local nodes accross the federation. By 'relevant' I mean tables that are
+# generated when triggering a udf execution accross several local nodes. By 'pointers'
+# I mean mapping between local nodes and table names and the aim is to hide the underline
+# complexity from the algorithm flow and exposing a single 'local node table' object that
+# stores in the background pointers to several tables in several local nodes.
+class LocalNodesTable(LocalNodesData):
+    def __init__(self, nodes_tables: Dict["LocalNode", NodeTable]):
```

Comment:
there is no "LocalNode" class

Reply:
The problem is that `algorithm_executor_node_data_objects` will import from `algorithm_executor_nodes` and the opposite, creating a circular import. It's not a matter of dependencied, it's only for type annotations, so I didn't merge the 2 files into one.

Reply:
the fact that a circular import will be created when you import something that is actually used in the module is a sign that something is not right.

Reply:
Not that it actually matters now, but there was no circular dependency on this part before. Nevertheless, one always has to fix a circular dependency as it is most likely a hint to problem(s) with the underlying design.

Reply:
On master:
a) https://github.com/madgik/MIP-Engine/blob/b68e995d93b2f4e51e498e62901eee0ae26691ce/mipengine/controller/algorithm_executor.py#L275
b) https://github.com/madgik/MIP-Engine/blob/b68e995d93b2f4e51e498e62901eee0ae26691ce/mipengine/controller/algorithm_executor.py#L240

I don't disagree that there is a problem, I am just saying that it was not introduced on this PR. The `_Node` and NodeData objects in general have circular dependencies. The `_Node` is using/returning data objects and the NodeData objects have the `_Node` in their initialization.

Reply:
The NodeData class is a new class introduced in this pr, so I do not understand your argument about the circular dependency being already there. 
Also the idea with the _INode interface and the _Node implementation of _INode was that they represent an instance of each "physical" node we have at the other end of the system. 
The idea with the _INodeTable interface and the _LocalNodeTable and _GlobalNodeTable implementation of _INodeTable was that they represent a db table on the global and the local node's db respectively.
Now, the new concepts of NodeData, AlgoExecData, LocalNodesData, GlobalNodeData is not clear to me what they actually respresent.

Reply:
After looking into it I found these:
Before this PR the `_Node` class needed the "INodeTable":
```
class _Node(_INode):
    .....
    def get_run_udf_query(
        self, command_id: str, func_name: str, positional_args: List["_INodeTable"]
    ) -> Tuple[str, str]:
    .....
```

and the `_LocalNodeTable` needed the `_Node`:
```
class _LocalNodeTable(INodeTable):
    def __init__(self, nodes_tables: Dict[_Node, TableName]):
        self._nodes_tables = nodes_tables
        self._validate_matching_table_names(list(self._nodes_tables.values()))
    .....
```

This was a bug though since the `get_run_udf_query` didn't receive as input a List["_INodeTable"] so when that was removed the circular importation problem could be fixed, my mistake, didn't remember how this all started.

I fixed it in the next PR: https://github.com/madgik/MIP-Engine/pull/175

```
Also the idea with the _INode interface and the _Node implementation of _INode was that they represent an instance of each "physical" node we have at the other end of the system.
The idea with the _INodeTable interface and the _LocalNodeTable and _GlobalNodeTable implementation of _INodeTable was that they represent a db table on the global and the local node's db respectively.
```
Nothing has changed on the logic of these.

```
Now, the new concepts of NodeData, AlgoExecData, LocalNodesData, GlobalNodeData is not clear to me what they actually represent.
```
The `NodeData` is the mother class of `NodeTable` (previous TableName) and `NodeSMPCTables`, nothing else.
The `AlgoExecData` is the mother class of `LocalNodesData` and `GlobalNodeData`, nothing else.
`LocalNodesData` is the mother class of `LocalNodesTable` (previously _LocalNodeTable) and `LocalNodesSMPCTables`, nothing else.
`GlobalNodeData` is the mother class of `GlobalNodeTable` (previously _GlobalNodeTable) and `GlobalNodeSMPCTables`, nothing else.

These classes were created to form a hierarchy, and type annotating, so that you know what to expect, as input/return in methods.

Does it make sense?